tool/tctl/sso/tester/command.go

```diff
@@ -0,0 +1,260 @@
+		types.KindGithubConnector: handleGithubConnector,
+	}
+
+	cmd.GetDiagInfoFields = map[string]func(diag *types.SSODiagnosticInfo, debug bool) []DiagnosticInfoField{
+		types.KindGithubConnector: getGithubDiagInfoFields,
+	}
+}
+
+func (cmd *SSOTestCommand) ssoTestCommand(c auth.ClientI) error {
+	reader := os.Stdin
+	if cmd.connectorFileName != "" {
+		f, err := utils.OpenFile(cmd.connectorFileName)
+		if err != nil {
+			return trace.Wrap(err, "could not open connector spec file %v", cmd.connectorFileName)
+		}
+		defer f.Close()
+		reader = f
+	}
+
+	decoder := kyaml.NewYAMLOrJSONDecoder(reader, defaults.LookaheadBufSize)
+	for {
+		var raw services.UnknownResource
+		err := decoder.Decode(&raw)
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return trace.Wrap(err, "Unable to load resource. Make sure the file is in correct format.")
+		}
+
+		handler, ok := cmd.Handlers[raw.Kind]
+		if !ok {
+			var kinds []string
+			for kind := range cmd.Handlers {
+				kinds = append(kinds, kind)
+			}
+			return trace.BadParameter("Resources of type %q are not supported. Supported kinds: %v", raw.Kind, kinds)
+		}
+
+		requestInfo, err := handler(c, raw.Raw)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// note: loginErr is processed further down.
+		loginResponse, loginErr := cmd.runSSOLoginFlow(context.TODO(), raw.Kind, c, requestInfo.Config)
+
+		if requestInfo.RequestCreateErr != nil {
+			return trace.BadParameter("Failed to create auth request. Check the auth connector definition for errors. Error: %v", requestInfo.RequestCreateErr)
+		}
+
+		info, infoErr := c.GetSSODiagnosticInfo(context.TODO(), raw.Kind, requestInfo.RequestID)
+
+		err = cmd.reportLoginResult(raw.Kind, info, infoErr, loginResponse, loginErr)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+}
+
+// TryRun is executed after the CLI parsing is done. The command must
+// determine if selectedCommand belongs to it and return match=true
+func (cmd *SSOTestCommand) TryRun(selectedCommand string, c auth.ClientI) (match bool, err error) {
+	if selectedCommand == cmd.ssoTestCmd.FullCommand() {
+		return true, cmd.ssoTestCommand(c)
+	}
+	return false, nil
+}
+
+type AuthRequestInfo struct {
+	Config           *client.RedirectorConfig
+	RequestID        string
+	RequestCreateErr error
+}
+
+func (cmd *SSOTestCommand) runSSOLoginFlow(ctx context.Context, protocol string, c auth.ClientI, config *client.RedirectorConfig) (*auth.SSHLoginResponse, error) {
+	key, err := client.NewKey()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	proxies, err := c.GetProxies()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if len(proxies) == 0 {
+		return nil, trace.BadParameter("cluster has no proxies.")
+	}
+
+	cfg := client.MakeDefaultConfig()
+	cfg.WebProxyAddr = proxies[0].GetPublicAddr()
+
+	tc, err := client.NewClient(cfg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```

Comment:
nit: extract  to a separate function

Reply:
Which one? I see no natural way to split `runSSOLoginFlow`.

Reply:
I was think about something like
```go
func runSSOLoginFlow() {
  key, err := client.NewKey()
  if err != nil {
           return nil, trace.Wrap(err)
  }
  tc, errr := cmd.createTeleportClient(...)
  if err != nil {
          return nil, trace.Wrap(err)
  }
  return return client.SSHAgentSSOLogin{
    ...
  }
}
```
 but I'm leaving it up to you.

Reply:
There would be no other uses for `createTeleportClient` beside the code here, so I think I'll leave it as is, at least for now.